File: php/tvmaze_sdk.php
<?php
declare(strict_types=1);

// Tvmaze SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

class TvmazeSDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new TvmazeUtility();
        $this->_utility = $utility;

        $config = TvmazeConfig::make_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features from config.
        $feature_opts = TvmazeHelpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $items = Struct::items($feature_opts);
            if ($items) {
                foreach ($items as $item) {
                    $fname = $item[0];
                    $fopts = TvmazeHelpers::to_map($item[1]);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, TvmazeFeatures::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return TvmazeUtility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = TvmazeHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = TvmazeHelpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = TvmazeHelpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new TvmazeSpec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    public function direct(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = TvmazeHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = TvmazeHelpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }


    private $_aka = null;

    // Idiomatic facade: $client->aka()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Aka() (PHP method
    // names are case-insensitive).
    public function aka($data = null)
    {
        require_once __DIR__ . '/entity/aka_entity.php';
        if ($data === null) {
            if ($this->_aka === null) {
                $this->_aka = new AkaEntity($this, null);
            }
            return $this->_aka;
        }
        return new AkaEntity($this, $data);
    }


    private $_alternate_list = null;

    // Idiomatic facade: $client->alternate_list()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias AlternateList() (PHP method
    // names are case-insensitive).
    public function alternate_list($data = null)
    {
        require_once __DIR__ . '/entity/alternate_list_entity.php';
        if ($data === null) {
            if ($this->_alternate_list === null) {
                $this->_alternate_list = new AlternateListEntity($this, null);
            }
            return $this->_alternate_list;
        }
        return new AlternateListEntity($this, $data);
    }


    private $_cast = null;

    // Idiomatic facade: $client->cast()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Cast() (PHP method
    // names are case-insensitive).
    public function cast($data = null)
    {
        require_once __DIR__ . '/entity/cast_entity.php';
        if ($data === null) {
            if ($this->_cast === null) {
                $this->_cast = new CastEntity($this, null);
            }
            return $this->_cast;
        }
        return new CastEntity($this, $data);
    }


    private $_cast_credit = null;

    // Idiomatic facade: $client->cast_credit()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias CastCredit() (PHP method
    // names are case-insensitive).
    public function cast_credit($data = null)
    {
        require_once __DIR__ . '/entity/cast_credit_entity.php';
        if ($data === null) {
            if ($this->_cast_credit === null) {
                $this->_cast_credit = new CastCreditEntity($this, null);
            }
            return $this->_cast_credit;
        }
        return new CastCreditEntity($this, $data);
    }


    private $_cast_member = null;

    // Idiomatic facade: $client->cast_member()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias CastMember() (PHP method
    // names are case-insensitive).
    public function cast_member($data = null)
    {
        require_once __DIR__ . '/entity/cast_member_entity.php';
        if ($data === null) {
            if ($this->_cast_member === null) {
                $this->_cast_member = new CastMemberEntity($this, null);
            }
            return $this->_cast_member;
        }
        return new CastMemberEntity($this, $data);
    }


    private $_crew = null;

    // Idiomatic facade: $client->crew()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Crew() (PHP method
    // names are case-insensitive).
    public function crew($data = null)
    {
        require_once __DIR__ . '/entity/crew_entity.php';
        if ($data === null) {
            if ($this->_crew === null) {
                $this->_crew = new CrewEntity($this, null);
            }
            return $this->_crew;
        }
        return new CrewEntity($this, $data);
    }


    private $_crew_credit = null;

    // Idiomatic facade: $client->crew_credit()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias CrewCredit() (PHP method
    // names are case-insensitive).
    public function crew_credit($data = null)
    {
        require_once __DIR__ . '/entity/crew_credit_entity.php';
        if ($data === null) {
            if ($this->_crew_credit === null) {
                $this->_crew_credit = new CrewCreditEntity($this, null);
            }
            return $this->_crew_credit;
        }
        return new CrewCreditEntity($this, $data);
    }


    private $_crew_member = null;

    // Idiomatic facade: $client->crew_member()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias CrewMember() (PHP method
    // names are case-insensitive).
    public function crew_member($data = null)
    {
        require_once __DIR__ . '/entity/crew_member_entity.php';
        if ($data === null) {
            if ($this->_crew_member === null) {
                $this->_crew_member = new CrewMemberEntity($this, null);
            }
            return $this->_crew_member;
        }
        return new CrewMemberEntity($this, $data);
    }


    private $_episode = null;

    // Idiomatic facade: $client->episode()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Episode() (PHP method
    // names are case-insensitive).
    public function episode($data = null)
    {
        require_once __DIR__ . '/entity/episode_entity.php';
        if ($data === null) {
            if ($this->_episode === null) {
                $this->_episode = new EpisodeEntity($this, null);
            }
            return $this->_episode;
        }
        return new EpisodeEntity($this, $data);
    }


    private $_guest_cast_credit = null;

    // Idiomatic facade: $client->guest_cast_credit()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias GuestCastCredit() (PHP method
    // names are case-insensitive).
    public function guest_cast_credit($data = null)
    {
        require_once __DIR__ . '/entity/guest_cast_credit_entity.php';
        if ($data === null) {
            if ($this->_guest_cast_credit === null) {
                $this->_guest_cast_credit = new GuestCastCreditEntity($this, null);
            }
            return $this->_guest_cast_credit;
        }
        return new GuestCastCreditEntity($this, $data);
    }


    private $_image = null;

    // Idiomatic facade: $client->image()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Image() (PHP method
    // names are case-insensitive).
    public function image($data = null)
    {
        require_once __DIR__ . '/entity/image_entity.php';
        if ($data === null) {
            if ($this->_image === null) {
                $this->_image = new ImageEntity($this, null);
            }
            return $this->_image;
        }
        return new ImageEntity($this, $data);
    }


    private $_person = null;

    // Idiomatic facade: $client->person()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Person() (PHP method
    // names are case-insensitive).
    public function person($data = null)
    {
        require_once __DIR__ . '/entity/person_entity.php';
        if ($data === null) {
            if ($this->_person === null) {
                $this->_person = new PersonEntity($this, null);
            }
            return $this->_person;
        }
        return new PersonEntity($this, $data);
    }


    private $_schedule = null;

    // Idiomatic facade: $client->schedule()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Schedule() (PHP method
    // names are case-insensitive).
    public function schedule($data = null)
    {
        require_once __DIR__ . '/entity/schedule_entity.php';
        if ($data === null) {
            if ($this->_schedule === null) {
                $this->_schedule = new ScheduleEntity($this, null);
            }
            return $this->_schedule;
        }
        return new ScheduleEntity($this, $data);
    }


    private $_scheduled_episode = null;

    // Idiomatic facade: $client->scheduled_episode()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias ScheduledEpisode() (PHP method
    // names are case-insensitive).
    public function scheduled_episode($data = null)
    {
        require_once __DIR__ . '/entity/scheduled_episode_entity.php';
        if ($data === null) {
            if ($this->_scheduled_episode === null) {
                $this->_scheduled_episode = new ScheduledEpisodeEntity($this, null);
            }
            return $this->_scheduled_episode;
        }
        return new ScheduledEpisodeEntity($this, $data);
    }


    private $_search = null;

    // Idiomatic facade: $client->search()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Search() (PHP method
    // names are case-insensitive).
    public function search($data = null)
    {
        require_once __DIR__ . '/entity/search_entity.php';
        if ($data === null) {
            if ($this->_search === null) {
                $this->_search = new SearchEntity($this, null);
            }
            return $this->_search;
        }
        return new SearchEntity($this, $data);
    }


    private $_season = null;

    // Idiomatic facade: $client->season()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Season() (PHP method
    // names are case-insensitive).
    public function season($data = null)
    {
        require_once __DIR__ . '/entity/season_entity.php';
        if ($data === null) {
            if ($this->_season === null) {
                $this->_season = new SeasonEntity($this, null);
            }
            return $this->_season;
        }
        return new SeasonEntity($this, $data);
    }


    private $_show = null;

    // Idiomatic facade: $client->show()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Show() (PHP method
    // names are case-insensitive).
    public function show($data = null)
    {
        require_once __DIR__ . '/entity/show_entity.php';
        if ($data === null) {
            if ($this->_show === null) {
                $this->_show = new ShowEntity($this, null);
            }
            return $this->_show;
        }
        return new ShowEntity($this, $data);
    }


    private $_update = null;

    // Idiomatic facade: $client->update()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Update() (PHP method
    // names are case-insensitive).
    public function update($data = null)
    {
        require_once __DIR__ . '/entity/update_entity.php';
        if ($data === null) {
            if ($this->_update === null) {
                $this->_update = new UpdateEntity($this, null);
            }
            return $this->_update;
        }
        return new UpdateEntity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new TvmazeSDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}

File: lua/tvmaze_sdk.lua
-- Tvmaze SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local TvmazeSDK = {}
TvmazeSDK.__index = TvmazeSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

TvmazeSDK._make_feature = _make_feature


function TvmazeSDK.new(options)
  local self = setmetatable({}, TvmazeSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features in the resolved order (make_options puts an explicit list
  -- order first, else defaults to test-first). Ordering matters: the `test`
  -- feature installs the base mock transport and the transport features
  -- (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
  -- must be added before them to sit at the base of the chain.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local featureorder = vs.getpath(self.options, "__derived__.featureorder")
    if type(featureorder) == "table" then
      for _, fname in ipairs(featureorder) do
        local fopts = helpers.to_map(feature_opts[fname])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function TvmazeSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function TvmazeSDK:get_utility()
  return Utility.copy(self._utility)
end


function TvmazeSDK:get_root_ctx()
  return self._rootctx
end


function TvmazeSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function TvmazeSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:Aka():list() / client:Aka():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:Aka(data)
  local EntityMod = require("entity.aka_entity")
  if data == nil then
    if self._aka == nil then
      self._aka = EntityMod.new(self, nil)
    end
    return self._aka
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:AlternateList():list() / client:AlternateList():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:AlternateList(data)
  local EntityMod = require("entity.alternate_list_entity")
  if data == nil then
    if self._alternate_list == nil then
      self._alternate_list = EntityMod.new(self, nil)
    end
    return self._alternate_list
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Cast():list() / client:Cast():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:Cast(data)
  local EntityMod = require("entity.cast_entity")
  if data == nil then
    if self._cast == nil then
      self._cast = EntityMod.new(self, nil)
    end
    return self._cast
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:CastCredit():list() / client:CastCredit():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:CastCredit(data)
  local EntityMod = require("entity.cast_credit_entity")
  if data == nil then
    if self._cast_credit == nil then
      self._cast_credit = EntityMod.new(self, nil)
    end
    return self._cast_credit
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:CastMember():list() / client:CastMember():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:CastMember(data)
  local EntityMod = require("entity.cast_member_entity")
  if data == nil then
    if self._cast_member == nil then
      self._cast_member = EntityMod.new(self, nil)
    end
    return self._cast_member
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Crew():list() / client:Crew():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:Crew(data)
  local EntityMod = require("entity.crew_entity")
  if data == nil then
    if self._crew == nil then
      self._crew = EntityMod.new(self, nil)
    end
    return self._crew
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:CrewCredit():list() / client:CrewCredit():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:CrewCredit(data)
  local EntityMod = require("entity.crew_credit_entity")
  if data == nil then
    if self._crew_credit == nil then
      self._crew_credit = EntityMod.new(self, nil)
    end
    return self._crew_credit
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:CrewMember():list() / client:CrewMember():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:CrewMember(data)
  local EntityMod = require("entity.crew_member_entity")
  if data == nil then
    if self._crew_member == nil then
      self._crew_member = EntityMod.new(self, nil)
    end
    return self._crew_member
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Episode():list() / client:Episode():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:Episode(data)
  local EntityMod = require("entity.episode_entity")
  if data == nil then
    if self._episode == nil then
      self._episode = EntityMod.new(self, nil)
    end
    return self._episode
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:GuestCastCredit():list() / client:GuestCastCredit():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:GuestCastCredit(data)
  local EntityMod = require("entity.guest_cast_credit_entity")
  if data == nil then
    if self._guest_cast_credit == nil then
      self._guest_cast_credit = EntityMod.new(self, nil)
    end
    return self._guest_cast_credit
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Image():list() / client:Image():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:Image(data)
  local EntityMod = require("entity.image_entity")
  if data == nil then
    if self._image == nil then
      self._image = EntityMod.new(self, nil)
    end
    return self._image
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Person():list() / client:Person():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:Person(data)
  local EntityMod = require("entity.person_entity")
  if data == nil then
    if self._person == nil then
      self._person = EntityMod.new(self, nil)
    end
    return self._person
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Schedule():list() / client:Schedule():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:Schedule(data)
  local EntityMod = require("entity.schedule_entity")
  if data == nil then
    if self._schedule == nil then
      self._schedule = EntityMod.new(self, nil)
    end
    return self._schedule
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:ScheduledEpisode():list() / client:ScheduledEpisode():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:ScheduledEpisode(data)
  local EntityMod = require("entity.scheduled_episode_entity")
  if data == nil then
    if self._scheduled_episode == nil then
      self._scheduled_episode = EntityMod.new(self, nil)
    end
    return self._scheduled_episode
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Search():list() / client:Search():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:Search(data)
  local EntityMod = require("entity.search_entity")
  if data == nil then
    if self._search == nil then
      self._search = EntityMod.new(self, nil)
    end
    return self._search
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Season():list() / client:Season():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:Season(data)
  local EntityMod = require("entity.season_entity")
  if data == nil then
    if self._season == nil then
      self._season = EntityMod.new(self, nil)
    end
    return self._season
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Show():list() / client:Show():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:Show(data)
  local EntityMod = require("entity.show_entity")
  if data == nil then
    if self._show == nil then
      self._show = EntityMod.new(self, nil)
    end
    return self._show
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Update():list() / client:Update():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function TvmazeSDK:Update(data)
  local EntityMod = require("entity.update_entity")
  if data == nil then
    if self._update == nil then
      self._update = EntityMod.new(self, nil)
    end
    return self._update
  end
  return EntityMod.new(self, data)
end




function TvmazeSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = TvmazeSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return TvmazeSDK
